### PR TITLE
[WIP][Common] The abstraction of Incremental Snapshot Framework

### PR DESCRIPTION
**Pull request by YesOrNo828**
Add a new module(flink-connector-base) which contain the common abstraction of incremental snapshot framework. So other db connectors could implement the Incremental Snapshot algorithm more easier. #724

### CODE REVIEW

File: flink-connector-base/src/main/java/com/ververica/cdc/connectors/base/source/dialect/Dialect.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	DataType fromDbzColumn(Column splitColumn);
+
+	/**
+	 * fetch current offset from the database
+	 * e.g. query Mysql binary logs by query <code>SHOW MASTER STATUS</code>.
+	 *
+	 * @param sourceConfig a basic source configuration.
+	 * @return current offset of the database.
+	 */
+	Offset createCurrentOffset(SourceConfig sourceConfig);
```
Comment (leonardBang) [inline flink-connector-base/src/main/java/com/ververica/cdc/connectors/base/source/dialect/Dialect.java:121]: how about `displayCurrentOffset ` ?

File: flink-connector-base/src/main/java/com/ververica/cdc/connectors/base/source/dialect/Dialect.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+						.getLogicalType();
+	}
+
+	/**
+	 * Get a corresponding Flink data type from a debezium {@link Column}
+	 *
+	 * @param splitColumn dbz split column.
+	 * @return flink data type
+	 */
+	DataType fromDbzColumn(Column splitColumn);
```
Comment (leonardBang) [inline flink-connector-base/src/main/java/com/ververica/cdc/connectors/base/source/dialect/Dialect.java:112]: This method looks strange in interface, Not every connector depends on debezium, right ?

File: flink-connector-base/src/main/java/com/ververica/cdc/connectors/base/source/dialect/SnapshotEventDialect.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.ververica.cdc.connectors.base.source.dialect;
+
+/**
+ * Created by yexianxun@corp.netease.com on 2022/1/7.
+ */
+public abstract class SnapshotEventDialect implements Dialect {
```
Comment (leonardBang) [inline flink-connector-base/src/main/java/com/ververica/cdc/connectors/base/source/dialect/SnapshotEventDialect.java:6]: I think the `SnapshotEventDialect` and `StreamingEventDialect ` should have different behaviors, but they need to implements same methods of `Dialect`.

**Review by leonardBang — COMMENTED** [review]
Thanks @YesOrNo828 for the awesome work , I think we're in right way.
I left several comments, and I think we can introduce a moudlue/connector like `mysql-connecor-cdc-new` which depends on our new framework, this won't influence current connector and help to validate our design in POC.
BTW, we can rebase master where I fix a log issue that prevent we look LOG in IDE.

**Comment by YesOrNo828** [comment]
@leonardBang  Thanks for you reviewing. 
I have rebased my branch.
And adding new module to validate the new framework that idea is good to me. I will do more test work.

File: flink-connector-base/src/main/java/com/ververica/cdc/connectors/base/source/dialect/Dialect.java (hunk 1 above)
Comment (YesOrNo828) [reply flink-connector-base/src/main/java/com/ververica/cdc/connectors/base/source/dialect/Dialect.java:121]: lgtm

File: flink-connector-mysql-cdc-new/src/main/java/com/ververica/cdc/connectors/mysql/refactor/schema/MySqlSchema.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
+package com.ververica.cdc.connectors.mysql.refactor.schema;
```
Comment (GOODBOY008) [inline flink-connector-mysql-cdc-new/src/main/java/com/ververica/cdc/connectors/mysql/refactor/schema/MySqlSchema.java:1]: Add license.

Comment (GOODBOY008) [reply flink-connector-mysql-cdc-new/src/main/java/com/ververica/cdc/connectors/mysql/refactor/schema/MySqlSchema.java:1]: @YesOrNo828 You can set `File and Code Template` with Apache License in idea.

**Comment by leonardBang** [comment]
@YesOrNo828 @GOODBOY008 I help fix the code style and conflicts, we can review&develop base on current code base.

**Comment by yangrong688** [comment]
Hi @YesOrNo828, @leonardBang , does this feature can apply to postgres-cdc when it complete.

**Comment by leonardBang** [comment]
Thanks @YesOrNo828 for the great work, I make some improvement base on your work, I think it's ok to merge now as the first version of framework.

**Review by leonardBang — APPROVED** [review]
+1

**Comment by 1032851561** [comment]
> Hi @YesOrNo828, @leonardBang , does this feature can apply to postgres-cdc when it complete.

+1

**Pull request merged**